### ISSUE TITLE
Scene Hard Clip is off by +4dB

### ISSUE THREAD

**Issue by mynameismuhl** [surge-synthesizer/surge#4117]
Linux/64bit/Reaper
Version:	1.9.main.99908055
steps to reproduce:

- Turn off your speakers
In Surge choose :
- Global Hard Clip Disabled
- Scene Output level 0dB
- Scene A Hard Clip at 0dBFS
- choose latch to play sound
- pull gain all the way up
- meters in Reaper peak at +4dBFS

To test Hard Clip at +18dBFS
you must turn off automute in Reaper first
(Options - Preferences - Mute/Solo)
Meters will peak at +22dBFS

Global Hard Clip clips at correct values.

**Comment by mkruselj** [comment]
Hmmm, can confirm this in SPAN too. Curious.

**Comment by baconpaul** [comment]
Another great report thank you.

I wonder if it is something to do with clipping before we downsample from the oversample rate? Unlikely. I also notice the global amplitude comes after the scene clip. Wonder if that’s it also?

Can look before we ship 1.9 for sure though!

**Comment by mynameismuhl** [comment]
So i read into /doc(s)
still i don't understand the signal flow

- HPF sits post Scene Output? If I deactivate it in above scenario 
meters now peak at 3.1dB (resonance?).

- I can send individual Scene Outputs to the daw pre masterFX,
excluding FX Return but still Scene Output is not independent
from Global Volume. So GV is not a stereo bus fader but more a VCA Fader
controlling all outputs?

OR

- insert FX are post HPF so post Scene Output 
but applied when scene is individualy routed to daw 
so SceneOutput != SceneOutput 
?

any of this true?

Tl;dr: 
confused.by --routing -h

**Comment by mkruselj** [comment]
HPF is at the end of scene output before insert FX.

**Comment by mynameismuhl** [comment]
Thanks,
and sorry my comment was kind of a mess.
What i meant was:
Is HPF located post Scene-Output-Attenuation/clipping?
Because if yes you found aprox.1dB.
The remaining 3dB could then be a mono-summing/pan-law issue somewhere post clipping.
But just guessing.

**Comment by mkruselj** [comment]
HPF is before output attenuation/clipping (you can see it in the code before the line I posted above).

**Comment by baconpaul** [comment]
So I am wondering (but am still not sure) if this is just a mislabeled menu. Let me explain.

When I first coded the menu I labeled them as “Hardclip at +/-1” and “Hardclip at +/-8” because that is what they do. The labels were then changed even though the code was not.

If you go read Wikipedia on dbfs it tells you 

> The unit dB FS or dBFS is defined in AES Standard AES17-1998,[13] IEC 61606,[14] and ITU-T Recs. P.381[15] and P.382,[16] such that the RMS value of a full-scale sine wave is designated 0 dB FS. This means a full-scale square wave would have an RMS value of +3 dB FS.[17][18] This convention is used in Wolfson[19] and Cirrus Logic[20] digital microphone specs, etc.

So in your case, you turn on a hard clipper, drive something at infiiite gain, clip at +/-1. That’s gonna make a full scale square wave if the clip is right and that’s not going to be the RMS of a 1/-1 sine.

So a mire interesting test is: there is some gain on the latched sine where the scene clip makes no difference. Does reaper show 0dbfs at that point.

I’m going to run the synth outside of reaper and look at the actual floats that are coming out. If the floats coming out in this case are actually +/-1 I propose we fix this by reverting the labels to the original mechanical form. (Again I want to test this first).

**Comment by mynameismuhl** [comment]
Wouldn't the global clipper give the same results in that case?

The init saw peaks at -8.5dBFS (in Reaper)
With scene-clipper at 0dBFS + global disabled:
if i add 8.6dB gain (in Surge) the meters now read 0.1dBFS
and so on. This slows down after the addition of 11dB (in total)
until it reaches a halt at 4dBFS with 23.5dB gain being added.

If i disable the scene-clipper and choose "Global Hard Clip at 0dBFS"
repeating this test the meters never go above 0dBFS.

**Comment by mkruselj** [comment]
So, somehow scene clipping results in a different dBFS value for the same signal level? Hmmm...

**Comment by baconpaul** [comment]
No no ... a +-1 sine wave has a different dbfs than an +-1 square wave is the point

**Comment by baconpaul** [comment]
Ahh right sorry read comment
Yeah ok like I said a theory. 
I will print out some values this week

**Comment by mynameismuhl** [comment]
Cool (and good morning btw).
These are further things i saw while playing with this - (maybe) being related to this issue:

Global-clip also somehow "overwrites" scene-clip
the same way global-volume previously controlled scene-volume
which can only be seen by routing scene to daw.

I was looking into SurgeSynthesizer.cpp.
I have very little understanding of programming so i describe what i think i see:
Sceneout Hardclip appears twice 
once before lowcut-filter and once in a for-loop after global hardclip.
If i remove this last switch statement and recompile just to see what happens
i now have two independent clippers. 
(The original issue still exists after this removal).

And activating the filter adds 0.9dB to the already clipped signal 
which it shouldn't be able to do if it was pre-sceneout-clipper
but i think in devices being patched into another not in code
maybe that's the problem?

**Comment by mkruselj** [comment]
OK we should definitely hardclip AFTER the highpass at the end of the scene. That is a fairly easy fix.

You see sceneout being clipped twice, because first time it's clipped prior to hitting the insert effects, and second time it's clipped after insert effects before sending that to individual plugin outputs (outs 3/4 for scene A and 5/6 for scene B).

**Comment by mynameismuhl** [comment]
sceneout clipping after insert doesn't seem to happen here.
I can still ripup gain with an insert and go above 4dBFS with sceneout enabled

**Comment by baconpaul** [comment]
OK here is exactly what is happening

The handclip sets the signal to +/-1 in oversample space
The half rate filter rings a little bit
it then passes that ringy signal the high pass filter
which rings it up a lot

Here is one block, for instance

```
PROCESS
B4 : 1335.27 <-- pre-clip RMS
AF : 1 <-- post clip scene RMS
HB : 1.00006 <-- post clip post halfband scene RMS
HF0 : 1.13147 <-- post HPF rms
```

So if I take your example in reaper where I set the clip to 0dbfs and then *deactivate the HPF* (with RMB/deactivate) then my RMS in reaper when I set gain to +48db and clip the scene is +3 (maybe +3.1) 

If I place the scene clip AFTER the half band filter and after the HPF, though, I get a 0 / 18 consistently in Reaper.

And since the default is 18 so we shouldn't really be hitting it anyway, I'm happy to merge that if you are.

I'l push a branch now with that change and @mkruselj leave it up to you to tell me if you think it is the right fix.

**Comment by baconpaul** [comment]
OK PR #4147 I think fixes this

If you agree @mkruselj and have tested it, please press the merge button

**Comment by mynameismuhl** [comment]
Yes that fixed the headline issue.
Regarding the remaining (sub)issues that came up during this discussion
- global clipper overwrites scene clipper when extra output is used.
- scene clipper is supposed to clip post insert but doesn't

i'll send a pull request that attempts to fix these.
If you agree and accept it my confusion about the output routing will dissolve
